### ISSUE TITLE
High Priority Msgs Seems Not Always Processed Before Normal Ones in Blocking Actor

### ISSUE THREAD

**Issue by zzxx-husky** [actor-framework/actor-framework#948]
Suppose two blocking actors `A` and `B`, `A` sends two messages `m1` then `m2` to `B`, where `m1` has high priority and `m2` has normal priority. The correct semantic should be that `B` processes `m1` before processing `m2` becuase `A` sends `m1` before `m2` to `B`. But I found that it's possible that `B` processes `m2` before `m1`.

I guess the problem is how the blocking actor [polls messages from the urgent queue and the normal queue](https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/src/blocking_actor.cpp#L267):

```c++
mailbox_element_ptr blocking_actor::dequeue() {
  mailbox().flush_cache();
  await_data();
  mailbox().fetch_more();
  auto& qs = mailbox().queue().queues();
  auto result = get<mailbox_policy::urgent_queue_index>(qs).take_front();
  if (!result)
   result = get<mailbox_policy::normal_queue_index>(qs).take_front();
  CAF_ASSERT(result != nullptr);
  return result;
}
```

Here the blocking actor first checks the urgent queue and then the normal queue. I think it's possible that when after the blocking actor checks the urgent queue and finds the urgent queue is empty, `m1` is put into the urgent queue and `m2` is put into the normal queue, then when the blocking actor checks the normal queue, it finds `m2` and processes it.

But I did not verify the guess above ... and I wonder if that guess is right?

**Comment by Neverlord** [comment]
> But I found that it's possible that B processes m2 before m1.

I haven't seen that behavior yet. Do you have a minimal examples reproducing this issue by chance?

> I guess the problem is how the blocking actor polls messages from the urgent queue and the normal queue:

In the snippet from `blocking_actor::dequeue`, `mailbox().fetch_more()` sorts the incoming messages into the queues. After that, we always ask the high priority queue first.

Just to make sure we're looking at the right function: do you actually use `dequeue` in your code? Because this code won't get called for `receive`, `receive_for`, etc. Using the receive function will take the path through [new_round](https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/src/blocking_actor.cpp#L255) on the mailbox in `receive_impl`.

**Comment by zzxx-husky** [comment]
> In the snippet from blocking_actor::dequeue, mailbox().fetch_more() sorts the incoming messages into the queues. After that, we always ask the high priority queue first.
>
> Just to make sure we're looking at the right function: do you actually use dequeue in your code? Because this code won't get called for receive, receive_for, etc. Using the receive function will take the path through new_round on the mailbox in receive_impl.

What I used is `receive_while`. I just went through the code path of `receive_while`. I found the code path finally ends at `receive_impl`. Sorry for pointing to the wrong snippet. 

I made the guess because I thought (although now it seems incorrect) that (1) each actor maintains two queues, one for normal messages and the other for urgent messages, (2) when a sender actor sends a message, the message is **directly** inserted into the normal (or urgent) queue of the receiver actor if the mesage is normal (or urgent). But now I found that both normal and urgent messages are actually put into the same `*queue*` of the mailbox, and the receiver actor gets messages from the mailbox and sorts messages of different priorities into different queues. Then, theoritically speaking, that guarantees the processing order of the messages (sent by the same sender) follows their creation order.

> But I found that it's possible that B processes m2 before m1.
However, that is still a problem to me. I tried to make some logs when `A` and `B` sends or receives messages. I found that `A` sent both `m1` and `m2`, but `B` only received `m2` and complained it did not receive any `m1`. Since message `m1` is idempotent, so I tried to send **two** `m1`s from A to B, one with normal priority and the other with high priority, then I found that `B` did not complain. I guess either (1) the `m1` with high priority is missed; or (2) `m1` with high priority is processed **after** `m2` because they are in different queues.

Sorry that I am chasing the deadline of a project at the moment T.T (that's also why I raised this issue without verification). May I hold this issue for some more time? I will catch up later.
